fuel-core/src/executor.rs

```diff
@@ -327,14 +354,105 @@ impl Executor {
+            coinbase_tx.outputs_mut().push(Output::coin(
+                self.config.block_producer.coinbase_recipient,
+                execution_data.coinbase,
+                AssetId::BASE,
+            ));
+            block.transactions[0] = coinbase_tx.clone().into();
+        }
+
+        coinbase_tx = self.check_coinbase(
+            block_height as Word,
+            coinbase_tx,
+            Some(execution_data.coinbase),
+        )?;
+        self.apply_coinbase(coinbase_tx, block, execution_data, block_db_transaction)?;
+
+        // Clean up unfinalized headers (block height + time + producer) that we inserted above
+        // for transaction execution.
+        block_db_transaction
+            .deref_mut()
+            .storage::<FuelBlocks>()
+            .remove(&Bytes32::zeroed())?;
+
+        Ok(data)
+    }
+
+    fn apply_coinbase(
+        &self,
+        coinbase_tx: Mint,
+        block: &PartialFuelBlock,
+        execution_data: &mut ExecutionData,
+        block_db_transaction: &mut DatabaseTransaction,
+    ) -> Result<(), Error> {
+        let coinbase_id = coinbase_tx.id();
+        self.persist_outputs(
+            *block.header.height(),
+            &coinbase_id,
+            block_db_transaction,
+            &[],
+            coinbase_tx.outputs(),
+        )?;
+        execution_data.tx_status.insert(
+            0,
+            (
+                coinbase_id,
+                TransactionStatus::Success {
+                    block_id: Default::default(),
+                    time: *block.header.time(),
+                    result: None,
+                },
+            ),
+        );
+        block_db_transaction
+            .deref_mut()
+            .storage::<Transactions>()
+            .insert(&coinbase_id, &coinbase_tx.into())?;
+        Ok(())
+    }
+
+    fn check_coinbase(
+        &self,
+        block_height: Word,
+        mint: Mint,
+        expected_amount: Option<Word>,
+    ) -> Result<Mint, Error> {
+        let checked_mint = mint
+            .into_checked(block_height, &self.config.chain_conf.transaction_parameters)?;
+
+        if checked_mint.transaction().tx_pointer().tx_index() != 0 {
+            return Err(Error::CoinbaseIsNotFirstTransaction)
         }
 
-        // TODO: Compare calculated coinbase with existing
-        // TODO: Insert at the first slot in production mode `Mint` transaction with right values.
-        // TODO: Apply coinbase transaction
+        if checked_mint.transaction().outputs().len() > 1 {
+            return Err(Error::CoinbaseSeveralOutputs)
+        }
+
+        if let Some(Output::Coin {
+            asset_id, amount, ..
+        }) = checked_mint.transaction().outputs().first()
+        {
+            if asset_id != &AssetId::BASE {
+                return Err(Error::CoinbaseOutputIsInvalid)
+            }
+
+            if let Some(expected_amount) = expected_amount {
+                if expected_amount != *amount {
```

Comment:
Maybe we shouldn't force the block producer to claim the full coinbase amount? They could have the option to burn the gas fees instead.

```suggestion
                if expected_amount < *amount {
```

Reply:
Hmm, I don't see how the block producer can use it to cheat, so we can allow it=)

With `!=`, we can know how much gas was spent, but with `<`, we can burn=) I'm okay with both variants, so you can commit your suggestion if you wish and merge PR(all tests are green, I checked locally)=)

Reply:
We can defer this for now